### ISSUE TITLE
Suggestions broken in some cases in v0.3.1

### ISSUE THREAD

**Issue by santiago0697** [HvyIndustries/crane#244]
Since the last update i cant use crane, here is the console output:
[Error - 10:13:42] Request textDocument/completion failed.
  Message: Request textDocument/completion failed with message: Cannot read property 'classes' of undefined
  Code: -32603 

Help me pls

**Comment by nevadascout** [comment]
Please can you supply some PHP code so we can reproduce this problem?

**Comment by cyberhorse** [comment]
Same problem. It happens in any PHP file, regardless of the code inside.

You asked for a sample, so here is one, but I don't think it will help:

```
<?php
class NewModel extends MY_Model
{
    public function __construct()
    {
        parent::__construct();
        $this->table = 'test';
    }
}
```

**Comment by ichiriac** [comment]
Hi @cyberhorse & @santiago0697,

I've also seen this error message before the 0.3.1 release made yesterday, but after the cache storage fix, I've didn't saw it again.

Could you check the extension version in order to sure you are using the latest version ?

Here the related issue : https://github.com/HvyIndustries/crane/issues/218

**Comment by cyberhorse** [comment]
Uninstalled and re-installed extension - same issue.

Which version you want me to check? Is there a cache I should try to delete? If yes, where?

**Comment by ichiriac** [comment]
Hi @cyberhorse,

If you have re-installed the version you might use the latest release. For anyone who may want to check their current version, go to the Extensions (CTRL+MAJ+X) and in the listing, you can see near the title "PHP Intellisense - Crane" a small number text with the version, example `0.3.1` for actual version.

By default we choose to disable debug informations in order to avoid verbose data. Can you go to your settings and set option `crane.debugMode` to `true`.

I've checked with your code sample, and also in various projects without being able to reproduce any error. I have 21 matches on the classes property so even if I can't reproduce the error, I need a stack or something that may explain the extension state in order to understand how to reproduce, or what may happened.

**Comment by cyberhorse** [comment]
Enabled crane.debugMode to true.
Same error, no additional output in the "Crane Language Server" log
In the "Crane Console" I get this:

```
[INFO] Checking the current version of Crane
[INFO] Watching these files: {**/*.php}
[INFO] Building project from cache file: **\AppData\Roaming/Crane/projects/fa591a6e9ff639e4acff950c579e0889/tree.cache
[INFO] Cache file successfully read
[INFO] Processing complete!
[INFO] File Changed: **File.php
[INFO] Processing complete!
```

**Comment by nevadascout** [comment]
@cyberhorse Please can you use the Command Palette to reindex the workspace, and then send us the output from the logs?

If you are on windows, press F1, then type "reindex" to find this option.

**Comment by cyberhorse** [comment]
I will do.

In the meantime, I tried opening a new window and starting a new file from scratch. I got this error:

```
[INFO] Checking the current version of Crane
[INFO] Watching these files: {**/*.php}
[ERROR] TypeError: Data must be a string or a buffer
    at TypeError (native)
    at Hash.update (crypto.js:73:16)
    at Cranefs.getProjectDir (C:\Users\ch\.vscode\extensions\HvyIndustries.crane-0.3.1\out\src\utils\Cranefs.js:74:60)
    at Promise (C:\Users\ch\.vscode\extensions\HvyIndustries.crane-0.3.1\out\src\utils\Cranefs.js:201:25)
    at Cranefs.createProjectFolder (C:\Users\ch\.vscode\extensions\HvyIndustries.crane-0.3.1\out\src\utils\Cranefs.js:200:16)
    at Promise (C:\Users\ch\.vscode\extensions\HvyIndustries.crane-0.3.1\out\src\utils\Cranefs.js:85:22)
    at Cranefs.createProjectDir (C:\Users\ch\.vscode\extensions\HvyIndustries.crane-0.3.1\out\src\utils\Cranefs.js:83:16)
    at Crane.processAllFilesInWorkspace (C:\Users\ch\.vscode\extensions\HvyIndustries.crane-0.3.1\out\src\crane.js:194:17)
    at Crane.doInit (C:\Users\ch\.vscode\extensions\HvyIndustries.crane-0.3.1\out\src\crane.js:156:18)
    at Crane.checkVersion.then.indexTriggered (C:\Users\ch\.vscode\extensions\HvyIndustries.crane-0.3.1\out\src\crane.js:33:18)
```

**Comment by cyberhorse** [comment]
Output of reindexing:

```
[INFO] Re-indexing PHP files in the workspace...
[INFO] Rebuilding the project files
[INFO] Preparing to parse 546 PHP source files...
[INFO] Preparing to parse workspace...
[INFO] Processing 92 stubs from C:\Users\ch\AppData\Roaming/Crane/phpstubs
[INFO] Stub files to process: 92
[INFO] 0 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/apc.php
[INFO] 1 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/bz2.php
[INFO] 2 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/apcu.php
[INFO] 3 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/basic.php
[INFO] 4 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/bcmath.php
[INFO] 5 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/calendar.php
[INFO] 6 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/couchbase.php
[INFO] 7 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/Core.php
[INFO] 8 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/ctype.php
[INFO] 9 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/curl.php
[INFO] 10 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/date.php
[INFO] 11 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/dba.php
[INFO] 12 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/enchant.php
[INFO] 13 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/dom.php
[INFO] 14 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/ereg.php
[INFO] 15 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/fileinfo.php
[INFO] 16 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/exif.php
[INFO] 17 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/filter.php
[INFO] 18 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/ftp.php
[INFO] 19 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/gd.php
[INFO] 20 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/geoip.php
[INFO] 21 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/gettext.php
[INFO] 22 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/gmp.php
[INFO] 23 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/gnupg.php
[INFO] 24 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/hash.php
[INFO] 25 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/imap.php
[INFO] 26 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/interbase.php
[INFO] 27 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/igbinary.php
[INFO] 28 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/iconv.php
[INFO] 29 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/imagick.php
[INFO] 30 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/json.php
[INFO] 31 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/ldap.php
[INFO] 32 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/libxml.php
[INFO] 33 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/mcrypt.php
[INFO] 34 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/mbstring.php
[INFO] 35 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/intl.php
[INFO] 36 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/memcache.php
[INFO] 37 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/memcached.php
[INFO] 38 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/mssql.php
[INFO] 39 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/mhash.php
[INFO] 40 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/mysql.php
[INFO] 41 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/mongo.php
[INFO] 42 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/mysqlnd.php
[INFO] 43 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/mysqli.php
[INFO] 44 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/oci8.php
[INFO] 45 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/OAuth.php
[INFO] 46 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/odbc.php
[INFO] 47 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/pcntl.php
[INFO] 48 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/openssl.php
[INFO] 49 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/pdo_dblib.php
[INFO] 50 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/pcre.php
[INFO] 51 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/PDO_Firebird.php
[INFO] 52 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/PDO_ODBC.php
[INFO] 53 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/pdo_mysql.php
[INFO] 54 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/pdo_sqlite.php
[INFO] 55 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/pdo_pgsql.php
[INFO] 56 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/PDO.php
[INFO] 57 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/pgsql.php
[INFO] 58 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/Phar.php
[INFO] 59 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/posix.php
[INFO] 60 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/recode.php
[INFO] 61 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/pspell.php
[INFO] 62 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/redis.php
[INFO] 63 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/readline.php
[INFO] 64 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/removed.php
[INFO] 65 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/Reflection.php
[INFO] 66 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/sasl.php
[INFO] 67 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/shmop.php
[INFO] 68 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/session.php
[INFO] 69 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/soap.php
[INFO] 70 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/SimpleXML.php
[INFO] 71 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/snmp.php
[INFO] 72 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/sockets.php
[INFO] 73 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/SPL.php
[INFO] 74 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/sqlite3.php
[INFO] 75 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/svn.php
[INFO] 76 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/sysvmsg.php
[INFO] 77 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/sysvsem.php
[INFO] 78 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/sysvshm.php
[INFO] 79 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/standard.php
[INFO] 80 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/tokenizer.php
[INFO] 81 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/wddx.php
[INFO] 82 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/tidy.php
[INFO] 83 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/xdebug.php
[INFO] 84 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/xml.php
[INFO] 85 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/xmlreader.php
[INFO] 86 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/xmlrpc.php
[INFO] 87 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/Zend-OPcache.php
[INFO] 88 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/xmlwriter.php
[INFO] 89 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/xsl.php
[INFO] 90 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/zip.php
[INFO] 91 Stub Processed: **/AppData/Roaming/Crane/phpstubs/crane-php-stubs-master/zlib.php
[INFO] Stubs parsing done!
[INFO] Workspace files to process: 546
[INFO] Parsed file 1 of 546 : \\server\projects\www\db.php
...
[INFO] Parsed file 546 of 546 : \\server\projects\system\database\drivers\pdo\subdrivers\pdo_sqlsrv_driver.php
[INFO] Workspace files have processed
[INFO] Processing complete!
[INFO] Project tree has been saved
```

**Comment by nevadascout** [comment]
Guessing from the `\\server` part, is your workspace on a mapped network drive?

**Comment by cyberhorse** [comment]
Yes.
When I tried opening a new window and creating a new file, still get the same problem though.

**Comment by nevadascout** [comment]
It looks like the `Cranefs` class does not properly handle network drives... I will investigate

**Comment by cyberhorse** [comment]
Thank you

**Comment by nevadascout** [comment]
Hey @cyberhorse if I send you a developer version of Crane with some extra debugging code in it, would you be able to install it for me and send me back the console output?

**Comment by cyberhorse** [comment]
Sure, happy to

**Comment by nevadascout** [comment]
I just found this which may be related: https://github.com/Microsoft/vscode/issues/5652

I have attached a dev version of Crane that writes out the workspace root path to the debug log when it starts up. Hopefully this will be enough to solve this issue.

You can install this version by doing the following:

1. Uninstall the current version of Crane in the extensions pane
2. Restart vscode
3. In the extensions pane click the three dots at the top right and select "Install from VSIX" at the bottom of the menu that appears, then select the attached `.vsix` file.
4. Reload vscode when prompted

[crane-0.3.2-alpha.vsix.zip](https://github.com/HvyIndustries/crane/files/829153/crane-0.3.2-alpha.vsix.zip)

**Comment by cyberhorse** [comment]
I think the autocomplete works with this version?! It doesn't always trigger automatically.
Also, there seems to be a bug, where the cursor loses focus from the current editor completely (not exactly sure what it focuses on, but I have to click back to where my cursor was before I can continue typing)

Crane Console output with the alpha version:

```
[INFO] Checking the current version of Crane
[INFO] Watching these files: {**/*.php}
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Building project from cache file: **\AppData\Roaming/Crane/projects/fa591a6e9ff639e4acff950c579e0889/tree.cache
[INFO] Workspace root path: \\server\project
[INFO] Cache file successfully read
[INFO] Processing complete!
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] File Changed: \\server\project\File.php
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] File Changed: \\server\project\File.php
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Processing complete!
[INFO] Processing complete!
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] File Changed: \\server\project\File.php
[INFO] Processing complete!
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] File Changed: \\server\project\File.php
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Processing complete!
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] File Changed: \\server\project\File.php
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] Workspace root path: \\server\project
[INFO] File Changed: \\server\project\File.php
[INFO] Processing complete!
[INFO] Processing complete!
```

**Comment by nevadascout** [comment]
That's very odd because I haven't specifically changed anything to fix this!

If it doesn't trigger automatically, make sure you have the 24/7 intellisense enabled. It should offer suggestions after key characters including `$`, `>`, `.`, etc

I assume that it opens the console when it logs out the debug information and that loses focus on the main editor. Do you still keep losing focus if you disable debug mode and try to type then?

**Comment by cyberhorse** [comment]
Disabling debug seems to keep the focus in. What's the 24/7 intellisense
option/plugin you mentioned?!

**Comment by nevadascout** [comment]
24/7 intellisense is a vs code feature that provides code suggestions as you are typing (all the time). The setting is `"editor.quickSuggestions": true` if you wanted to check you had it enabled.

Would you be able to make a short screen recording showing the intellisense not working?

**Comment by cyberhorse** [comment]
I would love to in principle, but won't be able to find the time for a while. 
I've recently (since after our last conversation) installed a CodeIgniter code completion, and that gives me some suggestions, but I don't get any suggestions from the current file/class for example.

**Comment by nevadascout** [comment]
Thanks, I'll keep investigating this issue!

**Comment by vdoria** [comment]
Got the same issue (0.3.3)? My problem started with 0.3.1 as well. I deactivated Crane and tried it again with 0.3.3... still not working :-/

Crane Language Server:
> [Error - 10:48:45] Request textDocument/completion failed.
  Message: Request textDocument/completion failed with message: Cannot read property 'classes' of undefined
  Code: -32603
> [Error - 10:48:50] Request textDocument/completion failed.
  Message: Request textDocument/completion failed with message: Cannot read property 'classes' of undefined
  Code: -32603

It happens on every file (just start typing $ for instance

I've tried to reindex but it gives me an error: 
> Running the contributed command:'crane.rebuildSources' failed.

I have the stubs installed (if that matters). I tried uninstalling it from GUI but reinstalling does not fix the issue. Still can't reindex due to above error.

**Comment by nevadascout** [comment]
Hi @vdoria thanks for letting us know!

Can you show me what the full filesystem path to your workspace folder is?

**Comment by nevadascout** [comment]
This may be fixed by the fix for #257, but I'll leave this open awaiting further information from @vdoria.

**Comment by cyberhorse** [comment]
Still an issue in 0.3.4
Also, I think duplicate of #274
